tasks/patch_server.rb

```diff
@@ -433,6 +433,56 @@ def run_pre_post_patching_script(scriptpath, pre_or_post, starttime, log = nil)
 end
 log.info "Apply only security patches set to #{security_only}"
 
+# Is there a package_list parameter? If security_only=true then package_list is ignored
+package_list_param = []
+if params['package_list'] && security_only == false
+
+  # Must be an array
+  if !params['package_list'].is_a?(Array)
+    err('106', 'pe_patch/params', 'Invalid parameter for package_list', starttime)
+  end
+
+  # Check for unsafe content
+  if params['package_list'].any? { |pkg| pkg =~ %r{[\$\|\/;`&]} }
+    err('107', 'pe_patch/package_list', 'Unsafe content in package_list', starttime)
+  end
+
+  # Install will fail if we try to update a package (or install a KB) that has no update available.
+  # Fail early/gracefully by checking each package/KB in package_list against the available updates.
+  #   - For Windows, each KB article ID must be a number and must exist in missing_update_kbs.
+  #   - For Linux, each package must exist in package_updates. Note that on RedHat package_updates
+  #     entries include architecture (e.g. 'foo-libs.x86_64') so factor this into the check.
+  if facts['values']['os']['family'] == 'windows'
+    missing_update_kbs = facts['values']['pe_patch']['missing_update_kbs']
+    params['package_list'].each do |kb_article_id|
+      if kb_article_id.to_i == 0
```

Comment:
Sorry, not quite sure I understand this; I think this says that we will error if `kb_article_id` is 0 with the message `KB ID is not a number: 0`. Am I missing something? I might be, I am very new to this repo!

Reply:
I think the intent is to check to see if it is a number at all.  Maybe `kb_article_id.to_i.to_s == kb_article_id` would be a better way to ensure it really is a number?

Reply:
The intent is to check if `kb_article_id` is a number, I'll change it per @jonathannewman suggestion. Thanks!

Reply:
`kb_article_id.to_i` returns `0` if `kb_article_id` does not contain a valid number:
`"123".to_i => 123`
`"abc".to_i => 0`

However (I hadn't realised):
`"123abc".to_i => 123`

So converting back to a string and comparing with the original is a better approach:
`"123abc".to_i.to_s == "123abc" => false`
`"123456".to_i.to_s == "123456" => true`

Reply:
Ruby is the gift that keeps on giving ... me headaches. :)